classes/metric/concurrent_users_metrics.php

```diff
@@ -0,0 +1,146 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace tool_cloudmetrics\metric;
+
+/**
+ * Metric class for users concurrently active.
+ *
+ * @package    metric_foobar
+ * @author     Marc-Alexandre Ghaly <marc-alexandreghaly@catalyst-ca.net>
+ * @copyright  2022, Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class concurrent_users_metrics extends builtin {
+    public $sample;
+    public $interval;
+    public $mintimestamp;
+    public $maxtimestamp;
+    public $sameconfig;
+
+    /**
+     * The metric's name.
+     *
+     * @return string
+     */
+    public function get_name(): string {
+        return 'concurrentusers';
+    }
+
+    /**
+     * The metric type.
+     *
+     * @return int
+     */
+    public function get_type(): int {
+        return manager::TYPE_GAUGE;
+    }
+
+    /**
+     * Retrieves the metric.
+     *
+     * @param  $backfilledrecord If data is backfilled, record is passed here.
+     * @return metric_item
+     */
+    public function get_metric_item($record = null): metric_item {
+        if (is_null($record)) {
+            global $DB;
+            $now = time();
+            $users = $DB->count_records_select('logstore_standard_log', 'timecreated > ?', [$now - $this->get_time_window()]);
+            return new metric_item($this->get_name(), $now, $users, $this);
+        }
+        return new metric_item($this->get_name(), $record->time, $record->value, $this);
+    }
+
+    /**
+     * Metric's ability to be backfilled.
+     *
+     * @return bool
+     */
+    public function is_backfillable(): bool {
+        return true;
+    }
+
+    /**
+     * Returns records for backfilled metric.
+     *
+     * @param array $data Array of data specifying timeframe and sample to backfill.
+     * @return array|null $metricitems Array of metrics items.
+     */
+    public function get_saved_metric_data(array $data): ?array {
+        global $DB;
+
+        $secondsinterval = [
+            MINSECS * 5,
+            HOURSECS,
+            DAYSECS,
+            WEEKSECS,
+            WEEKSECS * 4,
+            YEARSECS
+        ];
+        $sampleoptions = [
+            10,
+            100,
+            1000,
+            10000,
+            100000
+        ];
+
+        if (!empty($data)) {
+            $interval = $secondsinterval[$data['interval']] ?? MINSECS * 5;
+            $sample = $sampleoptions[$data['sample']] ?? 10;
+        } else {
+            // If form is empty we do not want to backfill anything.
+            return null;
+        }
+
+        $sql = "SELECT floor(timecreated / $interval ) * $interval AS time,
+                       COUNT(DISTINCT userid) as value
+                  FROM {logstore_standard_log}";
```

Comment:
We should not be querying directly from logstore_standard_log, there is an api for querying the logs instead:

https://github.com/moodle/moodle/blob/master/admin/tool/log/classes/log/manager.php#L93-L102